### DISCUSSION TITLE
Add current dateTime to function input query

### DISCUSSION THREAD

**Discussion by rbarreca** [Shopify/function-examples#416] [Feedback: General Improvements]
https://github.com/Shopify/function-examples/discussions/215 shipped date and comparisons, but no dateTime. @nickwesselman mentioned this is due to function run caching.

We need (and [others report needing](https://github.com/Shopify/function-examples/discussions/215)) the current dateTime. Here's our use-case: 

1. Merchant delivers local groceries and lets customers specify a delivery date before they get to checkout. This is set as a cart attribute as  `specifiedDeliveryDate`.
2. Merchant needs to close off ordering the day before delivery at 8:00 pm local time to prep for deliveries the following day.
3. Merchant wants a checkout validation function that prevents checkout if the current time is after 8:00 pm the day before the `specifiedDeliveryDate`.

### Actual behavior
There is no `dateTime`, only `date`, nor do any comparison functions work to compare the current dateTime to a variable. It currently only allows comparison to a metafield value, but we can't update a metafield value to be the 8:00 pm the day before the Cart attribute's `specifiedDeliveryDate`. 

To solve my use-case, I did a workaround using a combination of `date` and `timeAfter` but had to hardcode the cutoff time. Eventually, we'll want to have different cutoff times per delivery day and doing a dynamic cutoff of hour granularity would get a bit hairy as you'd have to add 24 `afterX...` fields.

```
shop {
  localTime {
    date
    after8pm: timeAfter(time: "20:00:00")
  }
}
```
### Expected behavior
Shopify adds a `dateTime` field to `LocalTime` that can be used in validations in a function run (e.g. compare it against a Cart attribute or Shop metafield that holds a hash like `{ "2024-01-04": "2024-01-03T20:00:00", ...}`).

**Comment by Adham-Mamedov** [comment]
Getting the exact time is critical to some custom logic and is an absolutely necessary feature. Example: Time the function needs to be active can be set to a half an hour or 45 minutes, due to store special offer campaign. Currently that is not possible to control.

**Reply by nickwesselman** [reply]
Can you explain more why `timeBefore`/`timeAfter` doesn't accomplish this?

**Reply by Adham-Mamedov** [reply]
I want to let user decide the exact time (Ex. 10:10 -- 12:35), so I don't know in advance what `timeBefore` and `timeAfter` to pass in graphQL query. The approach that rbarreca has used with `after8pm` cannot be used as I would need to have different options per minute (24 * 60), which is too big query.

**Reply by nickwesselman** [reply]
Would using [input query variables](https://shopify.dev/docs/apps/functions/input-output/variables-queries) solve this for you to pass in the time?

**Reply by rbarreca** [reply]
@nickwesselman are you imagining two variables in a JSON metafield like this passed into their function?

### Metafield JSON value
`{ "promoTimeStart": "16:30", "promoTimeEnd": "17:15" }`

### Input query

```
...
  timeBefore(time: $promoTimeStart),
  timeAfter(time: $promoTimeEnd)
...
```

**Reply by nickwesselman** [reply]
Yep exactly.

**Reply by nickwesselman** [reply]
Or even better, `timeBetween`:

https://shopify.dev/docs/api/functions/reference/product-discounts/graphql/common-objects/localtime

**Reply by Adham-Mamedov** [reply]
Thanks for suggestions, but there could be infinite time constraints. I am reusing cart-transform function for multiple campaigns. Each campaign can set its own date time settings. So it is not possible to define every set of timeStart and timeEnd in json and process in graphql query, without knowing the exact number of variables coming. I could predefine 10 or 20 variables ($promoTimeStart_X), but number of campaigns can go above 20. Although it works, that would be a bit of limitation.

**Reply by nickwesselman** [reply]
Our long-term recommended approach for this would probably be to separate the campaigns into multiple transforms / function instances. There are restrictions on the number of them at the moment, but we aim to expand this in the future.

**Comment by nickwesselman** [comment]
> To solve my use-case, I did a workaround using a combination of date and timeAfter but had to hardcode the cutoff time.

Is there a reason you can't use [input query variables](https://shopify.dev/docs/apps/functions/input-output/variables-queries) for this?

**Reply by rbarreca** [reply]
> Is there a reason you can't use [input query variables](https://shopify.dev/docs/apps/functions/input-output/variables-queries) for this?

We eventually want to do something fancier than the snippet I posted in the description. For the simplified `after8pm` example, we plan on moving to input query variables so that value comes from a metafield. That makes the hardcoded simplified example a little cleaner.

However, the fancier thing we want to do is to have a different cutoff time per delivery datetime to be able to do same-day delivery and handle weird holiday surges, etc. For example, Jan 12 9:00 am delivery window has cutoff of Jan 11 8pm but Jan 12 1:00 pm window has a cutoff of Jan 12 10:00 am, then on Thanksgiving we have an irregular cutoff for one specific day. If we tried to use input query variables for that use-case, we could put a JSON hash into a Shop-owned metafield like `{ "2024-01-12T09:00:00": "2024-01-11T20:00:00", ...}` but the GraphQL comparison functions wouldn't let us find a single value from a key to use in the comparison—it wants the whole metafield value. 

One follow-up: do the `timeAfter` functions evaluate in real-time every second or are those run on some wider granularity like every X minutes?

**Reply by nickwesselman** [reply]
`timeAfter` is evaluated every time the function input is calculated, prior to executing the function.

Would hour-level time accuracy in an input query allow you to fulfill the majority of your use cases? To the second or minute would really render us unable to cache function results.

**Reply by rbarreca** [reply]
For our use case, hour-level accuracy would be totally fine!

**Reply by manglestealth** [reply]
@nickwesselman Have you incorporated the current hour into the input for future planning?

**Reply by nickwesselman** [reply]
@manglestealth This is not planned work yet, can you add a comment describing your use case?

**Reply by manglestealth** [reply]
@nickwesselman 
Our use case is as follows: 
set a discount, then every X hours from its start time, the discount decreases by Y%. Therefore, during the execution of functions, we need to obtain the time difference from the current time to the start time. 
Currently, our solution is to use multiple `time0:timeAfter(time:"00:00:00")`. 
However, this solution has several issues: 
1. It can only be accurate to the hour level. If the start time of the discount is not at a whole hour, it is impossible to accurately determine whether an interval of time has passed. 
2. If using an hourly precision, it would require 24 timeAfter queries, nearly filling up the calculated query cost. If we need to query some metafields, the calculated query cost will exceed the maximum value of 30 in the limit. Therefore, in practice, we cannot even achieve hourly precision.

**Reply by nishikawa-nobuyuki** [reply]
@nickwesselman 
Even in my case, it is difficult to implement using timeAfter or timeBefore and requires datetime.

I am currently implementing the coming soon feature in my application.
In the meta-field of the product, the customer sets a time and the time is compared to the current date and time to determine if the product is available for sale.

I would like to use the cart and checkout Validation API for this, but the current spec makes it difficult to achieve.

This is because when we run.graphql, there is no information for the product metafield. We will only be executing run.graphql to get the time set by the customer from the meta field of the product.

Therefore, it is difficult to implement this application here with the current timeAfter and timeBefore specifications, and dateTime is required.

**Reply by saito-eiji** [reply]
I am having the same problem.

**Reply by fukumoto-taichi** [reply]
I'm facing the same issue. I would be very happy if the issue could be resolved.

**Reply by nickwesselman** [reply]
@nishikawa-nobuyuki Thanks for the use case description. It's not ideal but a workaround would be to put the product ID and availability time on a metafield on the validation object instead.

**Reply by km585** [reply]
I also need to use a current-time object in similar case [here](https://github.com/Shopify/function-examples/discussions/416#discussioncomment-8754565).

I understand that executing the function, which compares two different date-time objects , in the shopify backend and pass the result to our custom app is fast and accurate.   
But it is simply inconvenient that validation function cannot have a current time object.

If shopify really don't want us to have a datetime object inside a validation function, it would be nice if we could execute[ dateTimeAfter function](https://shopify.dev/docs/api/functions/reference/cart-checkout-validation/graphql/common-objects/localtime#:~:text=dateTimeAfter%20(Boolean!)) via api within validation function.

I really hope it will be possible for us to implement this type of validation soon.

**Reply by hasvtar** [reply]
@nickwesselman Hey Nick, I'm OK with implementing the workaround you mentioned using metafields (storing some date in a metafield) and then evaluate if that time has passed or not in our Shopify Function. 

However, my only worry is that I store the time along with a timezone in the metafield and I evaluate against it - what happens if there is a change in the stores timezone? The metafield value holding timezone becomes outdated and no longer accurate.

I'm wondering if there's any way to combat this by directly querying the stores timezone somehow within the function? I couldn't see any in the offical docs around getting the timezone of the store specifically: https://shopify.dev/docs/api/functions/reference/cart-checkout-validation/graphql/common-objects/localtime

Basically, I want my function to "activate" a field at the same time as every other function user (timezone agnostic) and to do this, I'd need to know the timezone of the shop.

**Reply by timfuhrmann** [reply]
> @nishikawa-nobuyuki Thanks for the use case description. It's not ideal but a workaround would be to put the product ID and availability time on a metafield on the validation object instead.

I have thought about moving the configuration of a launch date into our app where the information is written to the function to read as query input variables. As soon as there is more than one product to validate though, this solution is not sufficient, is it?

**Reply by nickwesselman** [reply]
@timfuhrmann You could use multiple, separate validations in this case

**Comment by fabregas4** [comment]
+1 on this. Have a need to be able to block checkout if items were added to cart more than X minutes ago. In the majority of cases X will be a short time-frame such as 15-30 minutes.

Idea would be to:
- set time added to cart as a hidden line item property when a product is first added to cart
- pull this in Function graphql and compare it to the current time to see how much time has passed. if > X then block checkout and give a message to customer

**Reply by Perugia** [reply]
I really need this solution. In my case, some cart line items need to be deleted or prevented from checkout 72 hours after being added to the cart.

**Reply by a-broz** [reply]
+1 exact same on my end...

**Reply by andrewchen5678** [reply]
+1

**Comment by abaileyb** [comment]
hey @nickwesselman and team- 

Running into an issue with localTime. I've removed the variables I was passing in the Input to isolate the problem as much as possible. 

My super contrived input looks like this:
`query RunInput {
  shop {
    hour0: localTime {
      timeBetween(startTime: "00:00:00", endTime: "01:00:00")
    }
    hour1: localTime {
      timeBetween(startTime: "01:00:00", endTime: "02:00:00")
    }
    hour2: localTime {
      timeBetween(startTime: "02:00:00", endTime: "03:00:00")
    }
    hour3: localTime {
      timeBetween(startTime: "03:00:00", endTime: "04:00:00")
    }
    hour4: localTime {
      timeBetween(startTime: "04:00:00", endTime: "05:00:00")
    }
    hour5: localTime {
      timeBetween(startTime: "05:00:00", endTime: "06:00:00")
    }
    hour6: localTime {
      timeBetween(startTime: "06:00:00", endTime: "07:00:00")
    }
    hour7: localTime {
      timeBetween(startTime: "07:00:00", endTime: "08:00:00")
    }
    hour8: localTime {
      timeBetween(startTime: "08:00:00", endTime: "09:00:00")
    }
    hour9: localTime {
      timeBetween(startTime: "09:00:00", endTime: "10:00:00")
    }
    hour10: localTime {
      timeBetween(startTime: "10:00:00", endTime: "11:00:00")
    }
    hour11: localTime {
      timeBetween(startTime: "11:00:00", endTime: "12:00:00")
    }
    hour12: localTime {
      timeBetween(startTime: "12:00:00", endTime: "13:00:00")
    }
    hour13: localTime {
      timeBetween(startTime: "13:00:00", endTime: "14:00:00")
    }
    hour14: localTime {
      timeBetween(startTime: "14:00:00", endTime: "15:00:00")
    }
    hour15: localTime {
      timeBetween(startTime: "15:00:00", endTime: "16:00:00")
    }
    hour16: localTime {
      timeBetween(startTime: "16:00:00", endTime: "17:00:00")
    }
    hour17: localTime {
      timeBetween(startTime: "17:00:00", endTime: "18:00:00")
    }
    hour18: localTime {
      timeBetween(startTime: "18:00:00", endTime: "19:00:00")
    }
    hour19: localTime {
      timeBetween(startTime: "19:00:00", endTime: "20:00:00")
    }
    hour20: localTime {
      timeBetween(startTime: "20:00:00", endTime: "21:00:00")
    }
    hour21: localTime {
      timeBetween(startTime: "21:00:00", endTime: "22:00:00")
    }
    hour22: localTime {
      timeBetween(startTime: "22:00:00", endTime: "23:00:00")
    }
    hour23: localTime {
      timeBetween(startTime: "23:00:00", endTime: "23:59:59")
    }
  }
}` 

All of the hourN's evaluate to false when the function runs:
<img width="312" alt="Screenshot 2024-12-16 at 7 41 39 PM" src="https://github.com/user-attachments/assets/92f746b5-8de3-4541-8a41-5f7e237c00b3" />

any thoughts?

**Reply by nickwesselman** [reply]
Hi @abaileyb -- Sorry to redirect but we don't provide support via this discussion board. Could you post to the community forums?

https://community.shopify.dev/c/shopify-functions/15

**Comment by aaaaargZombies** [comment]
+1 :pray: 

This would be really useful when you need to  compare time with a line-items attribute `offer-available-until` or similar. At the moment the `timeAfter` query only operates on a single time, but we would need current time to compare with multiple values to provide more precise and useful feedback to customers if a single line-item fowls a validation rule.

**Comment by JoshMealing** [comment]
Adding another use case here as I think this is a very important feature that is definitely needed within the community.

Products can be placed on a 24 hour hold by customers. A timestamp is added to a product metafield of when that hold expires.

If the hold date is lower than the current date, the hold is expired and the checkout/cart should be invalid.

This is the current structure of the Cart validation function input using version 2025-1 of the API:
```graphql
query RunInput {
  cart {
    lines {
      merchandise {
      ... on ProductVariant {
          product {
            hold_date: metafield(namespace:"hold", key: "date") {
              value
            }
          }
        }
      }
    }
  }

  shop {
    localTime {
      date
      has_expired_hold_future_date: dateTimeAfter(dateTime: "2025-02-28T12:00:00") # returns false
      has_expired_hold_past_date: dateTimeAfter(dateTime: "2025-02-24T12:00:00") # returns true
    }
  }

}
```

The dateTimeAfter and dateTimeBefore functions won't work for our use case, as these require the DateTime being hardcoded within the function input. We need to be able to use the DateTime stored in the product metafield.

Our ideal scenario for this would be if Shopify is able to add the DateTime into the LocalTime Object. That way, we can perform the check within the Function logic.

```graphql
  shop {
    localTime {
      date
      dateTime
    }
  }
```
Outputs

```javascript
{
   shop: {
      localTime: {
         date: "2025-02-25",
         dateTime: "2025-02-25T12:00:00"
      }
   }
}
```

**Reply by LostArrows27** [reply]
+1 for this feature. I'm currently struggling on how to check if my custom scheduling for discount volume is valid or not. Since i can't access current date with time.

**Reply by JoshMealing** [reply]
I added a new comment down below with the workaround we have used. Hopefully it helps with your situation

**Reply by LostArrows27** [reply]
@JoshMealing thanks a lot, i will give your solution a try !

**Comment by pohnean** [comment]
+1 for this. We also need the current local time. Using timeAfter/timeBetween is not an option because we have multi date and times to compare. 

This is because we are squeezing hundreds of discount configurations into 1 discount function as a workaround for Shopify's 5 function limit per app. Therefore, we also can't use the native startTime and endTime on discount functions and have to rely on localDate.

**Reply by davejcameron** [reply]
You can now have up to 25 automatic discounts (a mix of out of the box and powered by Functions) active at the same time.

**Reply by pohnean** [reply]
> You can now have up to 25 automatic discounts (a mix of out of the box and powered by Functions) active at the same time.

25 is not enough though. Some of out customers have a lot of promotions going on.

**Comment by JoshMealing** [comment]
I saw a few notifications coming through for new comments on this thread so i thought I would share the workaround we are using.

```graphql
# src/run.graphql
query RunInput {
  cart {
    date: attribute(key: "date") {
      key
      value
    }
    lines {
      merchandise {
      ... on ProductVariant {
          product {
            id
            productType
            title
            hold_date: metafield(namespace:"diamond", key: "hold_date") {
              value
            }
          }
        }
      }
    }
  }
}

```

On each page load within the theme code we are manually setting a Cart Attribute with the current date as a Unix time stamp.

In the Function input gql, we can then reference that cart attribute.

In the function logic we are comparing the cart attribute date with the product matefield date.

```javascript
// src/run.js
const { lines, date } = input.cart;
const currentDateTime = new Date(date?.value ?? 0);

const productHasCurrentHold = (holdDate) =>  new Date(parseInt(holdDate)) > currentDateTime;
......
......
.......
if (!productHasCurrentHold(hold_date_metafield.value)) {
  return {
    localizedMessage: `Unfortunately, your hold on the ${product.title} has expired.`,
    target: '$.cart'
  }
}

```

There are some pretty obvious drawbacks with this method, but it will get the job done until a more robust solution can be implemented by Shopify.

Drawbacks:
- Customers can use the Cart AJAX or Storefront APIs to manipulate the date and bypass the hold
- The code needs to run on each page load. It can also be run as a customer tries to navigate to the checkout, but this might not work with some themes
- If a customer navigates directly to the checkout(eg: viewing a checkout abandon email) the timestamp will not be set, potentially leading to a bypassed hold

There may be some other drawbacks I haven't considered, but that's all I've got for now. I hope this helps you guys.
